fuzzers/frida_executable_libpng/README.md

```diff
@@ -0,0 +1,36 @@
+# Fuzzing libpng with frida as executale
+
+This folder contains an example fuzzer for libpng, using LLMP for fast multi-process fuzzing and crash detection.
+To show off crash detection, we added a ud2 instruction to the harness, edit harness.cc if you want a non-crashing example.
+It has been tested on Linux.
+
+## Build
+
+To build this example, run `cargo build --release` in this folder.
+This will call (the build.rs)[./build.rs], which in turn downloads a libpng archive from the web.
+Then, it will build (the C++ harness)[./harness.cc] and the instrumented `libpng`.
+Then, it will create frida fuzzer shared library in `./target/release/libfrida_fuzzer.so`.
+On unix platforms, you'll need [libc++](https://libcxx.llvm.org/) to build it.
+
+Alternatively you can run `cargo make run` and this command will automatically build and run the fuzzer
+
+### Build For Android
+When building for android using a cross-compiler, make sure you have a [_standalone toolchain_](https://developer.android.com/ndk/guides/standalone_toolchain), and then add the following:
+1. In the ~/.cargo/config file add a target with the correct cross-compiler toolchain name (in this case aarch64-linux-android, but names may vary)
+`[target.aarch64-linux-android]`
+`linker="aarch64-linux-android-clang"`
+2. add path to installed toolchain to PATH env variable.
+3. define CLANG_PATH and add target to the build command line:
+`CLANG_PATH=<path to installed toolchain>/bin/aarch64-linux-android-clang cargo -v build --release --target=aarch64-linux-android`
+
+## Run
+
+This example uses in-process-fuzzing, using the `launcher` feature, in combination with a Restarting Event Manager.
+This means running --cores each client will start itself again to listen for crashes and timeouts.
+By restarting the actual fuzzer, it can recover from these exit conditions.
+
+After building the libpng-harness, you can run `find . -name libpng-harness` to find the location of your harness, then run
+
+```
+LD_PRELOAD=./target/release/libfrida_fuzzer.so ./libpng-harness --in corpus -out out
```

Comment:
@domenukk 

It still does not work as expected

When fuzzer is executed, it will resulted with the following output

```
 LD_PRELOAD=target/release/libfrida_fuzzer.so ./libpng-harness  -i corpus -o out
spawning on cores: Cores { cmdline: "0", ids: [CoreId { id: 0 }] }
child spawned and bound to core 0
I am broker!!.
18011 PostFork
New connection: 127.0.0.1:33964/127.0.0.1:33964
[Broker      #0]  (GLOBAL) run time: 0h-0m-30s, clients: 0, corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
                  (CLIENT) corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
[Broker      #0]  (GLOBAL) run time: 0h-1m-0s, clients: 1, corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
                  (CLIENT) corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
[Broker      #0]  (GLOBAL) run time: 0h-1m-30s, clients: 1, corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
                  (CLIENT) corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
[Broker      #0]  (GLOBAL) run time: 0h-2m-0s, clients: 1, corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
                  (CLIENT) corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
```

It seems like client does not send corpus, objective state to broker.

I'm still trying to debug it, why it does not working. 

Let me know if you have some tips or trick for debugging in-process frida.

Reply:
```
```

Reply:
Try to run main from rust without fuzzer and see how it goes
```
--- a/fuzzers/frida_executable_libpng/src/lib.rs
+++ b/fuzzers/frida_executable_libpng/src/lib.rs
@@ -6,7 +6,8 @@ pub unsafe extern "C" fn __libc_start_main(
     _argc: isize,
     _argv:*const *const char
 ) {
-    unsafe {
+    main(_argc, _argv);
+    /*unsafe {
         fuzzer::lib(main);
-    }
+    }*/
 }
```

It seems like segvault was produce when `main` function exits and jump into `__libc_start_main`

```
gdb> set environment LD_PRELOAD ./target/release/libfrida_fuzzer.so
gdb> file libpng-harness
gdb> start 16 AAABBBCCCDDDEEEFFFHHHIIIJJJKKKLLLLMMM
gdb> ni
     0x402676 <main+262>       mov    eax, DWORD PTR [rbp-0xbc]
     0x40267c <main+268>       cmp    eax, 0x0
     0x40267f <main+271>       je     0x4026b9 <main(int,  char**)+329>
 →   0x402685 <main+277>       mov    DWORD PTR [rbp-0x4], 0x0
     0x40268c <main+284>       mov    DWORD PTR [rbp-0x50], 0x1
     0x402693 <main+291>       jmp    0x403036 <main(int,  char**)+2758>
     0x402698 <main+296>       mov    QWORD PTR [rbp-0x48], rax
     0x40269c <main+300>       mov    DWORD PTR [rbp-0x4c], edx
     0x40269f <main+303>       lea    rdi, [rbp-0x40]
gdb> ni
   0x7ffff5cbc151 <__libc_start_main+193> mov    rdi, QWORD PTR [rsp]
   0x7ffff5cbc155 <__libc_start_main+197> mov    rsi, QWORD PTR [rsp+0x8]
   0x7ffff5cbc15a <__libc_start_main+202> call   r14
 → 0x7ffff5cbc15d <__libc_start_main+205> add    rsp, 0x50
   0x7ffff5cbc161 <__libc_start_main+209> pop    rbx
   0x7ffff5cbc162 <__libc_start_main+210> pop    r14
   0x7ffff5cbc164 <__libc_start_main+212> pop    r15
   0x7ffff5cbc166 <__libc_start_main+214> ret    
   0x7ffff5cbc167                  nop    WORD PTR [rax+rax*1+0x0]

gdb> ni
     0x402366 <_start+22>      mov    rcx, 0x4246f0
     0x40236d <_start+29>      mov    rdi, 0x402570
     0x402374 <_start+36>      call   QWORD PTR [rip+0x2fc76]        # 0x431ff0
 →   0x40237a <_start+42>      hlt    
     0x40237b                  nop    DWORD PTR [rax+rax*1+0x0]
     0x402380 <_dl_relocate_static_pie+0> ret    
     0x402381                  nop    WORD PTR cs:[rax+rax*1+0x0]
     0x40238b                  nop    DWORD PTR [rax+rax*1+0x0]
     0x402390 <deregister_tm_clones+0> mov    eax, 0x4321b0
gdb> ni
Single stepping until exit from function _start,
which has no line number information.

Program received signal SIGSEGV, Segmentation fault.
0x000000000040237a in _start ()

```

Reply:
By adding exit after main function is called, fix the Segmentation fault issue

https://github.com/AFLplusplus/LibAFL/pull/1117/commits/359fef5e183cf5edd52288491e96e43291e24483#diff-2e0ead35a2f1c9e1f89c753c83783a7ae6213ed9b484b32961fbd948b1949607R23

But still the broker didnt get any feedback from client

Reply:
Are you setting up -F, -H, -l correctly?

in frida_libpng, you run the fuzzer with this
```
./frida_fuzzer -F LLVMFuzzerTestOneInput -H ./libpng-harness.so -l ./libpng-harness.so
```

F is the function to instrument. (in your case this is main() perhaps)
l is the additional libs to instrument.

So this means you need to tell the fuzzer what to instrument. else frida won't dynamically patch the code to instrument

Reply:
@tokatoka Thanks for the nudge!!! Indeed I forgot to add those flag. 

Here is the output when trying to execute it. It works
```
 LD_PRELOAD=./target/debug/libfrida_fuzzer.so ./libpng-harness -i corpus -o out -H ./libpng-harness  
spawning on cores: Cores { cmdline: "0", ids: [CoreId { id: 0 }] }
child spawned and bound to core 0
I am broker!!.
22104 PostFork
New connection: 127.0.0.1:44248/127.0.0.1:44248
[Stats       #1]  (GLOBAL) run time: 0h-0m-0s, clients: 2, corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000
                  (CLIENT) corpus: 0, objectives: 0, executions: 0, exec/sec: 0.000, edges: 972/65536 (1%)
[Testcase    #1]  (GLOBAL) run time: 0h-0m-0s, clients: 2, corpus: 1, objectives: 0, executions: 1, exec/sec: 0.000
                  (CLIENT) corpus: 1, objectives: 0, executions: 1, exec/sec: 0.000, edges: 972/65536 (1%)
[Stats       #1]  (GLOBAL) run time: 0h-0m-0s, clients: 2, corpus: 1, objectives: 0, executions: 1, exec/sec: 0.000
                  (CLIENT) corpus: 1, objectives: 0, executions: 1, exec/sec: 0.000, edges: 1150/65536 (1%)
[Testcase    #1]  (GLOBAL) run time: 0h-0m-0s, clients: 2, corpus: 2, objectives: 0, executions: 2, exec/sec: 0.000
                  (CLIENT) corpus: 2, objectives: 0, executions: 2, exec/sec: 0.000, edges: 1150/65536 (1%)

```